.gitlab-ci.yml

```diff
@@ -80,13 +76,42 @@ default:
   tags:
     - linux-docker
 
+.docker-env-only:                       &docker-env-only
+  only:
+    - master
+    - /^v[0-9]+\.[0-9]+.*$/        # i.e. v1.0, v2.1rc1
+    - schedules
+    - web
+    - /^[0-9]+$/                   # PRs
+
 .build-only:                       &build-only
   only:
     - master
     - /^v[0-9]+\.[0-9]+.*$/        # i.e. v1.0, v2.1rc1
     - /^pre-v[0-9]+\.[0-9]+-[0-9a-f]+$/
     - web
 
+.build-rules:                       &build-rules
+  rules:
+    - if: '$DEPLOY_TAG'
+      when: never
+    - if: $CI_COMMIT_REF_NAME=="master"
+      when: always
+    - if: $CI_PIPELINE_SOURCE=="web"
+      when: always
+    - if: $CI_COMMIT_REF_NAME=~ /^v[0-9]+\.[0-9]+.*$/
+      when: always
+    - if: $CI_COMMIT_REF_NAME=~ /^pre-v[0-9]+\.[0-9]+-[0-9a-f]+$/
```

Comment:
Hey, good job here! These two must be tested with the matching tag names. The problem is GitLab mirrors the branch #s instead of names, but should mirror the tag names. And tag names are prioritized over the branch name by `CI_COMMIT_REF_NAME` variable.
Also, did you check that this `&build-rules` works the same way as the previous `only` from `&docker-env`? It really seems like it should, but this is a dark part of the GitLab config.

Reply:
i tested on another repo. Seems working as desired on syntaxes (happy path). 
I need your judgement here.